File: src/main/java/world/ultravanilla/commands/SpectateCommand.java
package world.ultravanilla.commands;

import world.ultravanilla.AnarchyRegion;
import world.ultravanilla.UltraVanilla;
import world.ultravanilla.reference.LegacyColors;
import world.ultravanilla.reference.Palette;
import world.ultravanilla.reference.Users;
import world.ultravanilla.serializable.Position;

import net.md_5.bungee.api.ChatColor;

import org.bukkit.GameMode;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabExecutor;
import org.bukkit.entity.Player;
import world.ultravanilla.UltraVanilla;
import world.ultravanilla.reference.Users;
import world.ultravanilla.serializable.Position;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class SpectateCommand extends UltraCommand implements CommandExecutor {

    public static final ChatColor COLOR = ChatColor.WHITE;

    public SpectateCommand(UltraVanilla instance) {
        super(instance);
        this.color = COLOR;
    }

    @Override
    public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
        if (sender instanceof Player) {
            Player player = (Player) sender;

            UUID uuid = player.getUniqueId();

            if (Users.spectators.contains(uuid)) {
                player.teleport(((Position) plugin.getConfig().get("spawn")).getLocation());
                player.setGameMode(GameMode.SURVIVAL);
                Users.spectators.remove(player.getUniqueId());
                UltraVanilla.set(player, "spectator", false);
            } else {
                player.teleport(AnarchyRegion.center());
                player.setGameMode(GameMode.SPECTATOR);
                Users.spectators.add(player.getUniqueId());
                UltraVanilla.set(player, "spectator", true);
            }
        }

        return true;
    }

}

File: src/main/java/world/ultravanilla/commands/SpectateCommand.scala
package world.ultravanilla.commands

import net.md_5.bungee.api.ChatColor
import org.bukkit.GameMode
import org.bukkit.command.{Command, CommandExecutor, CommandSender}
import org.bukkit.entity.Player
import world.ultravanilla.reference.Users
import world.ultravanilla.serializable.Position
import world.ultravanilla.{AnarchyRegion, UltraVanilla}

object SpectateCommand {
    val COLOR = ChatColor.WHITE
}

class SpectateCommand(val instance: UltraVanilla) extends UltraCommand(instance) with CommandExecutor {
    this.color = SpectateCommand.COLOR

    override def onCommand(sender: CommandSender, command: Command, label: String, args: Array[String]): Boolean = {
        if (sender.isInstanceOf[Player]) {
            val player = sender.asInstanceOf[Player]
            val uuid = player.getUniqueId
            if (Users.spectators.contains(uuid)) {
                val spawn = plugin.getConfig.get("spawn").asInstanceOf[Position].getLocation

                player.setGameMode(GameMode.SURVIVAL)
                player.teleport(spawn)
                Users.spectators.remove(player.getUniqueId)
                UltraVanilla.set(player, "spectator", false)

                plugin.scheduleSyncTask(1) { () =>
                    player.setGameMode(GameMode.SURVIVAL)
                    player.teleport(spawn)
                }
            } else {
                player.setGameMode(GameMode.SPECTATOR)
                player.teleport(AnarchyRegion.center)
                Users.spectators.add(player.getUniqueId)
                UltraVanilla.set(player, "spectator", true)
            }
        }
        true
    }
}
